### ISSUE TITLE
Softmax activation function

### ISSUE THREAD

**Issue by wizard1989** [Dobiasd/frugally-deep#13]
Hello, @Dobiasd 

Current softmax function is applied to all values of tensor3.

For example, I paint the regions of a picture (300x300 pixels) with the SegNet-like network to denote the location of objects of 7 different types.
The last layer has input tensor3 with a shape (7, 300, 300).
Current softmax function applies exp() function to all 7x300x300.
Then all values are divided by the sum of all 7x300x300 values.

But the softmax function is needed to be applied along the channel dimension (as it's done in "keras") in order to classify each pixel type.

I've changed the softmax_layer code to support this.
The code is tested and presented below.
Please, consider adding the new code to the repository.

```
// Copyright 2016, Tobias Hermann.
// https://github.com/Dobiasd/frugally-deep
// Distributed under the MIT License.
// (See accompanying LICENSE file or at
//  https://opensource.org/licenses/MIT)

#pragma once

#include "fdeep/layers/layer.hpp"

namespace fdeep { namespace internal
{

class softmax_layer : public activation_layer
{
public:
    explicit softmax_layer(const std::string& name)
        : activation_layer(name)
    {
    }
protected:
    tensor3 transform_input(const tensor3& input) const override
    {
        // Softmax function is applied along the channel dimension.

        const auto ex = [this](float_type x) -> float_type
        {
            return std::exp(x);
        };

        // Get unnormalised values of exponent function.
        auto output = transform_tensor3(ex, input);

        for (size_t iRow = 0; iRow < input.shape().height_; ++iRow)
        {
            for (size_t jColumn = 0; jColumn < input.shape().width_; ++jColumn)
            {
                // Get the sum of unnormalised values for one pixel.
                // We are not using Kahan summation algorithm due to usually
                // small number of object classes.
                float_type sum = 0.0;
                for (size_t kClass = 0; kClass < input.shape().depth_; ++kClass)
                {
                    sum += output.get(kClass, iRow, jColumn);
                }
                // Divide the unnormalised values for one pixel by the sum.
                for (size_t kClass = 0; kClass < input.shape().depth_; ++kClass)
                {
                    output.set(kClass, iRow, jColumn, output.get(kClass, iRow, jColumn)/sum);
                }
            }
        }

        return output;

    }
};

} } // namespace fdeep, namespace internal
```

**Comment by Dobiasd** [comment]
Hi Valeriy,

thank you very much! I hope running into this problem with fdeep did not cause you too much trouble. Obviously I simply forgot to add [a test case](https://github.com/Dobiasd/frugally-deep/blob/master/keras_export/generate_test_models.py) for this usage of `softmax_layer` and thus did not mimic the behavior of Keras correctly.

Would you like to create a [PR](https://github.com/Dobiasd/frugally-deep/pulls) including your change, or would you prefer if I add it to the code base myself?

**Comment by wizard1989** [comment]
No problems. I've created PR.

**Comment by Dobiasd** [comment]
fixed with [PR15](https://github.com/Dobiasd/frugally-deep/pull/15)
